go/ai/tools.go

```diff
@@ -37,18 +36,51 @@ type Tool interface {
 	Action() action.Action
 	// RunRaw runs this tool using the provided raw map format data (JSON parsed
 	// as map[string]any).
-	RunRaw(ctx context.Context, input map[string]any) (any, error)
+	RunRaw(ctx context.Context, input any) (any, error)
+}
+
+// ToolInterruptError represents an intentional interruption of tool execution.
+type ToolInterruptError struct {
+	Metadata map[string]any
+}
+
+func (e *ToolInterruptError) Error() string {
+	return "tool execution interrupted"
+}
+
+// InterruptOptions provides configuration for tool interruption.
+type InterruptOptions struct {
+	Metadata map[string]any
 }
 
-// DefineTool defines a tool function.
-func DefineTool[In, Out any](r *registry.Registry, name, description string, fn func(ctx context.Context, input In) (Out, error)) *ToolDef[In, Out] {
+// ToolContext provides context and utility functions for tool execution.
+type ToolContext struct {
+	Context   context.Context
```

Comment:
is that... "allowed"??

Reply:
I mean, is it considered idiomatic in Go?
Reading the code below it definitely seems like a pretty clean approach, so I have no objections. Just checking, don't want to offend Go purists.

Reply:
Yes, apparently there are two idiomatic ways of doing this: 1) helper getter/setters for `context.Context` and 2) wrapping `context.Context` in another struct and passing that in place of it. It just depends. I may end up doing the former later but this is fine for now.